### DISCUSSION TITLE
The way that mkdocs handles creating HTML anchors for duplicate headings is less than ideal for my situation.

### DISCUSSION THREAD

**Discussion by rmscode** [mkdocs/mkdocs#3860] [Q&A] [answered]
I use mkdocs for note keeping as well as documentation. The way that I take my notes is from bottom to top (top being the most recent). Here's an example:

```markdown
# 2024

## November

### Th 11.07.2024

**1300** - Lunch was great!

---

**0800** - Breakfast was great!

### W 11.06.2024

**1300** - Lunch was bad!
```

I recently decided to format my time entries as headings so I could be more granular when linking back to specific notes.

```markdown
# 2024

## November

### Th 11.07.2024

#### 1300

Lunch was great!

---

#### 0800

Breakfast was great!

### W 11.06.2024

#### 1300

Lunch was bad!
```

That has proven to be less useful as I thought since when there are duplicate headings, mkdocs appends a number to the end of the HTML anchor. The number designations to the end of the HTML anchors are updated iteratively/sequentially. In the example above, the most recent `#### 1300` entry will be `2024notes.html#1300` and the next is `2024notes.html#1300_1`. If I were to add a third `#### 1300` entry in the coming days, that will become `2024notes.html#1300` and the rest will change sequentially to follow.

`2024notes.html#1300`     --becomes--> `2024notes.html#1300_1`
`2024notes.html#1300_1` --becomes--> `2024notes.html#1300_2`

I could add the date to my time entry headings, but that would be redundant and look a bit silly. This isn't a huge issue...I'm just looking for suggestions. Maybe there is a better way for me to take notes using mkdocs.

**Comment by Andre601** [comment, marked as answer]
You can use the `attr_list` extension to override the ID that would be set for the header like this:

```markdown
#### 1300 { #1300_1 }

Lorem Ipsum

#### 1300 { #1300_2 }
```
This would make the first header use `1300_1` and the second `1300_2`

**Reply by rmscode** [reply]
Thats a great suggestion, but I think it would be a nightmare for me to manage. I have hundreds if not thousands of individual note entries.

---

Edit: On second thought...maybe it wouldn't be that terrible. I could use regex find and replace to in vscode to quickly make all the necessary adjustments. Then, since I already use a custom keybinding to print my times, I could adjust that as well. Something like this:

```json
{
    "key": "ctrl+d t",
    "command": "editor.action.insertSnippet",
    "when": "editorTextFocus",
    "args": {
        "snippet": "#### $CURRENT_HOUR$CURRENT_MINUTE { #$CURRENT_MONTH.$CURRENT_DATE.$CURRENT_YEAR-$CURRENT_HOUR$CURRENT_MINUTE }"
    }
}
```

That would print something like `#### 1020 { #11072024-1020 }` for me 😎

---

Future readers may find this useful . . . I ended up using github copilot to assist me with creating a Python script to parse my existing markdown and make the necessary changes.

```python
import re

def process_markdown(file_path):
    with open(file_path, 'r', encoding='utf-8') as file:
        lines = file.readlines()

    date_pattern = re.compile(r'### \w+ (\d{2}\.\d{2}\.\d{4})')
    time_pattern = re.compile(r'#### (\d{4})')
    current_date = None

    for i, line in enumerate(lines):
        date_match = date_pattern.match(line)
        if date_match:
            current_date = date_match.group(1).replace('.', '')

        time_match = time_pattern.match(line)
        if time_match and current_date:
            time = time_match.group(1)
            lines[i] = f'#### {time} {{ #{current_date}-{time} }}\n'

    with open(file_path, 'w', encoding='utf-8') as file:
        file.writelines(lines)

# Replace '2024.md' with the path to your markdown file
process_markdown('2024.md')
```

**Comment by pawamoy** [comment]
@Andre601 provided a good answer. An alternative, if you'd prefer making this automatic, would be to write a Python-Markdown extension, using a tree processor, in order to update a heading's id depending on its parent headings.

**Reply by rmscode** [reply]
I thought of this too. I have been chipping away at learning python over the past several months. Maybe this would be a great first project for me. Thanks for the suggestions!
